pkg/storage/s3.go

```diff
@@ -1,56 +1,288 @@
 package storage
 
 import (
+	"bytes"
+	"io/ioutil"
+	"net/url"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/spf13/pflag"
+
 	"github.com/pingcap/errors"
 	"github.com/pingcap/kvproto/pkg/backup"
-	"github.com/spf13/pflag"
+)
+
+var (
+	sendCredential bool
 )
 
 const (
-	s3EndpointOption = "s3.endpoint"
-	s3RegionOption   = "s3.region"
+	s3EndpointOption     = "s3.endpoint"
+	s3RegionOption       = "s3.region"
+	s3StorageClassOption = "s3.storage-class"
+	s3SSEOption          = "s3.sse"
+	s3ACLOption          = "s3.acl"
+	s3ProviderOption     = "s3.provider"
+	notFound             = "NotFound"
+	// number of retries to make of operations
+	maxRetries = 3
 )
 
-// S3BackendOptions are options for configurating the S3 storage.
+// s3Handlers make it easy to inject test functions
+type s3Handlers interface {
+	HeadObject(*s3.HeadObjectInput) (*s3.HeadObjectOutput, error)
+	GetObject(*s3.GetObjectInput) (*s3.GetObjectOutput, error)
+	PutObject(*s3.PutObjectInput) (*s3.PutObjectOutput, error)
+	HeadBucket(*s3.HeadBucketInput) (*s3.HeadBucketOutput, error)
+	WaitUntilObjectExists(*s3.HeadObjectInput) error
+}
+
+// S3Storage info for s3 storage
+type S3Storage struct {
+	session *session.Session
+	svc     s3Handlers
+	options *backup.S3
+}
+
+// S3BackendOptions contains options for s3 storage
 type S3BackendOptions struct {
-	Endpoint string `json:"endpoint" toml:"endpoint"`
-	Region   string `json:"region" toml:"region"`
+	Endpoint              string `json:"endpoint" toml:"endpoint"`
+	Region                string `json:"region" toml:"region"`
+	StorageClass          string `json:"storage_class" toml:"storage_class"`
```

Comment:
```suggestion
	StorageClass          string `json:"storage-class" toml:"storage-class"`
```

Reply:
Backup proto uses `_`, https://github.com/pingcap/kvproto/blob/master/proto/backup.proto#L131-L143

Reply:
@SunRunAway Why unify with kvproto? We are not sending these options through gRPC. We use `-` in all config files.